### PR TITLE
Rebase of PR 3303 (sparse matrix norms)

### PR DESCRIPTION
**Pull request by argriffing**
Fixes merge conflicts that have accumulated over the last year.
https://github.com/scipy/scipy/pull/3303
~~The PR still needs work, especially with respect to `axis`-specific norms and complex-valued matrices.~~ The original authorship meta-data is intact.

### CODE REVIEW

**Comment by argriffing** [comment]
I've made some modifications, completely removing the `axis` argument, fixing the complex-valued code path, and increasing test coverage on the remaining code.  Perhaps a later PR could restore the `axis` argument for sparse matrix norms.

**Comment by argriffing** [comment]
The red X may have delayed comments for a couple of weeks, so when TravisCI gets around to style-checking my changes again this may be fixed.

File: scipy/sparse/linalg/_norm.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    """
+    Norm of a sparse matrix
+
+    This function is able to return one of seven different matrix norms,
+    depending on the value of the ``ord`` parameter.
+
+    Parameters
+    ----------
+    x : a sparse matrix
+        Input sparse matrix. If `axis` is None, `x` must be 1-D or 2-D
```
Comment (rgommers) [inline scipy/sparse/linalg/_norm.py]: One left-over mention of ``axis`` here.

**Comment by rgommers** [comment]
Commit 0dcf1b5769 doesn't look so dramatic, what was the problem with the ``axis`` argument exactly?

**Comment by argriffing** [comment]
@rgommers Maybe the `axis` could be added back in a later PR?  There is not a deep problem with `axis`, but there are several shallow issues that could potentially cause this PR to become stalled.

**Comment by rgommers** [comment]
All the other recently added commits (besides the ``axis`` one) look fine to me. @pv had already okay''d the rest it looks like. So seems good to merge, modulo the ``axis`` question.

File: scipy/sparse/linalg/_norm.py (hunk 1 above)
Comment (argriffing) [inline scipy/sparse/linalg/_norm.py]: Removed.

**Comment by rgommers** [comment]
@argriffing OK thanks. Let's give this a day or two for @pv or @perryism to comment, if no further comments let's merge as is. Might be good in that case to open a new issue after merging for adding ``axis`` back with a reference to the commit that removed it and listing the known issues with it.

**Comment by argriffing** [comment]
> Might be good in that case to open a new issue

I've added one here:
https://github.com/scipy/scipy/issues/4547

**Comment by ewmoore** [comment]
Given all of the norm related PRs currently in flight, what interface do we ultimately want scipy to provide for norms?  Do we want something like #4369?  If we do, does it make sense to only expose that kind of interface for sparse matrices instead of the `numpy.linalg.norm` interface in this PR?  I don't know that I have a strong opinion, but I think we should probably pick one or the other.

**Comment by argriffing** [comment]
@ewmoore given the relative slowness of the scipy development cycle especially for 'inclusion' as opposed to 'deletion' PRs, it could make more sense to just close that other PR.  Maybe it could be repurposed as a norm interface in Julia or in a hypothetical 'scidynd' or 'sciblaze'.

**Comment by rgommers** [comment]
gh-4369 is already closed now. I don't see a problem in @argriffing's proposal to merge this one and other norm-related PRs first, and then reconsider that PR. Given that this PR mirrors the current ``linalg`` one it's good to have this in either way I'd think.

**Comment by rgommers** [comment]
OK in it goes. Thanks all!

**Comment by ewmoore** [comment]
Ok. Seems good.  My concern was that we had a number of things in flight, without having any real plan.

**Pull request merged**